test/uri_internal_test.c

```diff
@@ -0,0 +1,522 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/safestack.h>
+
+#include "internal/uri.h"
+
+#include "e_os.h"
+
+struct test_data_st {
+    int good;
+    char *uri;
+    struct {
+        char *scheme;
+        int decoded_scheme;
+        char *authority;
+        int decoded_authority;
+        char *path;
+        int decoded_path;
+        char *query;
+        int decoded_query;
+        char *fragment;
+        int decoded_fragment;
+        char *user;
+        int decoded_user;
+        char *password;
+        int decoded_password;
+        char *host;
+        int decoded_host;
+        char *service;
+        int decoded_service;
+    } expected;
+};
+typedef struct test_data_st TEST_DATA;
+
+static void free_test_data(TEST_DATA *test_data)
+{
+    if (test_data == NULL)
+        return;
+    OPENSSL_free(test_data->uri);
+    OPENSSL_free(test_data->expected.scheme);
+    OPENSSL_free(test_data->expected.authority);
+    OPENSSL_free(test_data->expected.path);
+    OPENSSL_free(test_data->expected.query);
+    OPENSSL_free(test_data->expected.fragment);
+    OPENSSL_free(test_data->expected.user);
+    OPENSSL_free(test_data->expected.password);
+    OPENSSL_free(test_data->expected.host);
+    OPENSSL_free(test_data->expected.service);
+    OPENSSL_free(test_data);
+}
+
+static int parse_key_value(const char *line, char **key, char **value)
+{
+    size_t off = 0;
+
+    while(*line != '\n' && isspace(*line))
+        line++;
+
+    off = 0;
+    while(line[off] != '\n' && !isspace(line[off]) && line[off] != '=')
+        off++;
+
+    if ((*key = OPENSSL_strndup(line, off)) == NULL)
+        return 0;
+
+    line += off;
+    while(*line != '\n' && isspace(*line))
+        line++;
+
+    if (line[0] != '=') {
+        OPENSSL_free(*key);
+        return 0;
+    }
+    line++;
+    while(*line != '\n' && isspace(*line))
+        line++;
+
+    off = strlen(line);
+    while(off > 0 && isspace(line[off - 1]))
+        off--;
+
+    if ((*value = OPENSSL_strndup(line, off)) == NULL) {
+        OPENSSL_free(*key);
+        return 0;
+    }
+
+    return 1;
+}
+
+static int linecounter = 0;
+static TEST_DATA *read_test_data(FILE *f)
+{
+    TEST_DATA *test_data = NULL;
+    int errcount = 0;
+    int skip_stanza = 0;
+
+    while (1) {
+        char line[2048], *p = NULL, *key = NULL, *value = NULL;
+
+        if (fgets(line, sizeof(line), f) == NULL) {
+            if (ferror(f)) {
+                perror("reading tests");
+                free_test_data(test_data);
+                test_data = NULL;
+            }
+            break;
+        }
+
+        linecounter++;
+
+        for (p = line; *p != '\n' && isspace(*p); p++)
+            ;
+
+        if (*p == '\n') {
+            if (test_data != NULL)
+                break;
```

Comment:
`read_test_data()` reads one stanza per call.  Stanzas are delimited by empty lines.  So, if it finds an empty line and `test_data` is non-NULL, it means it's finished reading one stanza and must break out of the loop.